File: scripts/verify-review-protocol-v1-10-candidate.mjs
#!/usr/bin/env node

import fs from 'node:fs';
import { spawnSync } from 'node:child_process';

import {
  assertClosedObject,
  assertRepoRelativePath,
  canonical,
  contentRoot,
  pretty,
  readRegularFileNoFollow,
  sha256Bytes,
} from './b0-v8-core.mjs';
import {
  REVIEW_PROTOCOL_FINDINGS,
  REVIEW_ROLE_SLOTS,
  REVIEW_VALIDATOR_IDS,
  evaluateCurrentReviewProtocolState,
  evaluateReviewProtocolInput,
  makeReviewProtocolVector,
  runReviewProtocolMutationCampaign,
} from './review-protocol-v1-10-core.mjs';

const DATE = '2026-08-30';
const PACKAGE_DIR = `docs/planning/three-review-protocol-v1-10-g1-package-${DATE}`;
const REPORT_DIR = `docs/planning/three-review-protocol-v1-10-g1-detached-reports-${DATE}`;
const PATHS = Object.freeze({
  manifest: `${PACKAGE_DIR}/normative-package-manifest.json`, registry: `${PACKAGE_DIR}/normative-registry.json`, sourceIndex: `${PACKAGE_DIR}/frozen-source-index.json`, crosswalk: `${PACKAGE_DIR}/closure-crosswalk.json`, corpus: `${PACKAGE_DIR}/mutation-corpus.json`,
  reportA: `${REPORT_DIR}/qa-reader-a-report.json`, reportB: `${REPORT_DIR}/qa-reader-b-report.json`,
});

function runGit(args, encoding = 'utf8', maxBuffer = 4 * 1024 * 1024) {
  const result = spawnSync('git', args, { cwd: process.cwd(), encoding, maxBuffer });
  if (result.status !== 0) throw new Error(`git ${args.join(' ')} failed: ${encoding ? result.stderr.trim() : result.stderr.toString('utf8').trim()}`);
  return result.stdout;
}

function readGitBlob(commitOid, logicalPath, maxBytesExclusive) {
  assertRepoRelativePath(logicalPath);
  const entries = runGit(['ls-tree', '-z', '--full-tree', commitOid, '--', logicalPath], null).toString('utf8').split('\0').filter(Boolean);
  if (entries.length !== 1) throw new Error(`Git path must resolve exactly once: ${logicalPath}`);
  const match = /^(100644|100755) blob ([0-9a-f]{40,64})\t(.+)$/.exec(entries[0]);
  if (!match || match[3] !== logicalPath) throw new Error(`Git path is not an exact regular blob: ${logicalPath}`);
  const bytes = runGit(['cat-file', 'blob', match[2]], null, maxBytesExclusive);
  if (bytes.length >= maxBytesExclusive) throw new Error(`Git blob exceeds byte limit: ${logicalPath}`);
  return { byteLength: bytes.length, bytes, logicalPath, mode: Number.parseInt(match[1].slice(-3), 8), sha256: sha256Bytes(bytes) };
}

function readJson(repositoryRoot, logicalPath, maxBytes = 25 * 1024 * 1024) {
  const fact = readRegularFileNoFollow(repositoryRoot, logicalPath, maxBytes);
  return { fact, value: JSON.parse(fact.bytes.toString('utf8')) };
}

function validateManifest(repositoryRoot, manifest) {
  const keys = ['artifactClass', 'artifactId', 'generatedAt', 'maxMemberBytesExclusive', 'maxTotalBytesInclusive', 'memberCount', 'members', 'packageContentRoot', 'packageId', 'repositoryVisibility', 'schemaVersion', 'sourceCommit', 'totalBytes'];
  assertClosedObject(manifest, keys, 'manifest');
  if (manifest.repositoryVisibility !== 'PUBLIC' || manifest.schemaVersion !== 'MPRR-V1-10-PACKAGE-MANIFEST-V1' || !Array.isArray(manifest.members) || manifest.members.length !== 11 || manifest.memberCount !== 11) throw new Error('manifest invariant mismatch');
  const paths = new Set(); const hashes = new Set(); const roles = new Set(); let total = 0;
  manifest.members.forEach((member, index) => {
    assertClosedObject(member, ['bytes', 'logicalPath', 'ordinal', 'role', 'sha256'], `manifest.members[${index}]`);
    if (member.ordinal !== index + 1 || paths.has(member.logicalPath) || hashes.has(member.sha256) || roles.has(member.role)) throw new Error('manifest order or uniqueness mismatch');
    paths.add(member.logicalPath); hashes.add(member.sha256); roles.add(member.role);
    const fact = member.logicalPath.startsWith('scripts/') || member.logicalPath.startsWith('tests/') ? readGitBlob(manifest.sourceCommit, member.logicalPath, manifest.maxMemberBytesExclusive) : readRegularFileNoFollow(repositoryRoot, member.logicalPath, manifest.maxMemberBytesExclusive);
    if (fact.sha256 !== member.sha256 || fact.byteLength !== member.bytes) throw new Error(`manifest member drift: ${member.logicalPath}`);
    total += member.bytes;
  });
  if (!manifest.members.some((member) => member.role === 'MPRRV110-B0-CORE-DEPENDENCY' && member.logicalPath === 'scripts/b0-v8-core.mjs')) throw new Error('manifest missing transitive B0 core dependency');
  if (total !== manifest.totalBytes || total > manifest.maxTotalBytesInclusive) throw new Error('manifest total budget mismatch');
  const projection = { members: manifest.members, packageId: manifest.packageId, sourceCommit: manifest.sourceCommit };
  if (contentRoot('MPRR-V1-10-PACKAGE-CONTENT-V1', projection) !== manifest.packageContentRoot) throw new Error('manifest package root mismatch');
}

function validateRegistry(registry) {
  const keys = ['algorithmPolicy', 'artifactClass', 'artifactId', 'controlRows', 'currentState', 'externalEvidence', 'growthPolicy', 'ownerModel', 'receiptSchema', 'reportPolicy', 'repositoryVisibility', 'roleSlots', 'schemaVersion', 'validatorIds'];
  assertClosedObject(registry, keys, 'registry');
  if (registry.repositoryVisibility !== 'PUBLIC' || registry.schemaVersion !== 'MPRR-V1-10-NORMATIVE-REGISTRY-V1') throw new Error('registry invariant mismatch');
  if (canonical(registry.validatorIds) !== canonical(REVIEW_VALIDATOR_IDS) || canonical(registry.roleSlots) !== canonical(REVIEW_ROLE_SLOTS)) throw new Error('registry validator/role denominator mismatch');
  if (registry.ownerModel.workOwner !== 'Tal' || registry.ownerModel.ownerCount !== 1 || registry.ownerModel.primaryBackupModel !== 'REMOVED') throw new Error('registry owner model mismatch');
  if (canonical(registry.currentState) !== canonical(evaluateCurrentReviewProtocolState())) throw new Error('registry current state mismatch');
  if (registry.externalEvidence.externalClosureCount !== 0 || registry.externalEvidence.independentReviewerEvidencePresent !== false || registry.externalEvidence.operationalEvidencePresent !== false) throw new Error('registry external evidence overclaim');
  if (registry.algorithmPolicy.approvedAlgorithms.length !== 0 || registry.algorithmPolicy.keyGenerationPerformed !== false || registry.algorithmPolicy.selectionState !== 'UNSELECTED-PER-USE-APPROVAL-REQUIRED') throw new Error('registry cryptographic policy mismatch');
  if (registry.reportPolicy.safeDescriptorBoundAdapterPresent !== false || registry.reportPolicy.writesAllowed !== false) throw new Error('registry report adapter overclaim');
  if (registry.growthPolicy.globalBudgetPresent !== false || registry.growthPolicy.duplicateBytesMustBeDerived !== true || !registry.growthPolicy.admission.startsWith('DENIED')) throw new Error('registry growth policy mismatch');
  const { schemaRoot, ...schemaBase } = registry.receiptSchema;
  if (schemaBase.additionalProperties !== false || contentRoot('MPRR-V1-10-CLOSED-SCHEMA-V1', schemaBase) !== schemaRoot) throw new Error('registry receipt schema mismatch');
  if (!Array.isArray(registry.controlRows) || registry.controlRows.length !== 17) throw new Error('registry control denominator mismatch');
  registry.controlRows.forEach((row, index) => {
    assertClosedObject(row, ['closureStatus', 'controlId', 'externalEvidenceRequired', 'findingId', 'localStatus', 'severity', 'testId'], `control[${index}]`);
    if (row.findingId !== REVIEW_PROTOCOL_FINDINGS[index][0] || row.severity !== REVIEW_PROTOCOL_FINDINGS[index][1] || row.localStatus !== 'IMPLEMENTED-CANDIDATE' || row.closureStatus !== 'OPEN-PENDING-INDEPENDENT-EVIDENCE') throw new Error('registry control overclaim or identity mismatch');
  });
}

function validateSourceIndex(index, sourceCommit) {
  assertClosedObject(index, ['artifactClass', 'artifactId', 'repositoryVisibility', 'rows', 'schemaVersion', 'sourceCount', 'sourceSetRoot'], 'sourceIndex');
  if (index.repositoryVisibility !== 'PUBLIC' || !Array.isArray(index.rows) || index.rows.length !== 11 || index.sourceCount !== 11) throw new Error('source index denominator mismatch');
  const paths = new Set(); const hashes = new Set();
  index.rows.forEach((row, position) => {
    assertClosedObject(row, ['bytes', 'logicalPath', 'mode', 'ordinal', 'sha256'], `source[${position}]`);
    if (row.ordinal !== position + 1 || paths.has(row.logicalPath) || hashes.has(row.sha256)) throw new Error('source index order/uniqueness mismatch');
    paths.add(row.logicalPath); hashes.add(row.sha256);
    const fact = readGitBlob(sourceCommit, row.logicalPath, 50 * 1024 * 1024);
    if (fact.sha256 !== row.sha256 || fact.byteLength !== row.bytes || fact.mode !== row.mode) throw new Error(`source drift: ${row.logicalPath}`);
  });
  if (!index.rows.some((row) => row.logicalPath === 'scripts/b0-v8-core.mjs')) throw new Error('source index missing transitive B0 core dependency');
  if (contentRoot('MPRR-V1-10-SOURCE-SET-V1', index.rows) !== index.sourceSetRoot) throw new Error('source index root mismatch');
}

function validateCrosswalk(registry, crosswalk) {
  assertClosedObject(crosswalk, ['artifactClass', 'artifactId', 'closureCount', 'crosswalkRoot', 'findingCount', 'inheritedFindingCount', 'inheritedIndependentMechanicalClosureCount', 'repositoryVisibility', 'rows', 'schemaVersion'], 'crosswalk');
  if (crosswalk.findingCount !== 17 || crosswalk.closureCount !== 0 || crosswalk.inheritedFindingCount !== 40 || crosswalk.inheritedIndependentMechanicalClosureCount !== 1 || !Array.isArray(crosswalk.rows) || crosswalk.rows.length !== 17) throw new Error('crosswalk denominator mismatch');
  crosswalk.rows.forEach((row, index) => {
    const { ordinal, ...control } = row;
    if (ordinal !== index + 1 || canonical(control) !== canonical(registry.controlRows[index])) throw new Error('crosswalk row mismatch');
  });
  const { crosswalkRoot, ...base } = crosswalk;
  if (contentRoot('MPRR-V1-10-CROSSWALK-V1', base) !== crosswalkRoot) throw new Error('crosswalk root mismatch');
}

function validateCorpus(corpus) {
  assertClosedObject(corpus, ['artifactClass', 'artifactId', 'blockedCount', 'caseCount', 'cases', 'corpusRoot', 'repositoryVisibility', 'schemaVersion'], 'corpus');
  if (corpus.caseCount !== 17 || corpus.blockedCount !== 17 || !Array.isArray(corpus.cases) || corpus.cases.length !== 17) throw new Error('corpus denominator mismatch');
  corpus.cases.forEach((row, index) => {
    assertClosedObject(row, ['actual', 'findingId', 'mutationId', 'ordinal', 'target', 'testRoot'], `mutation[${index}]`);
    const { testRoot, ...base } = row;
    if (row.ordinal !== index + 1 || row.findingId !== REVIEW_PROTOCOL_FINDINGS[index][0] || row.actual !== 'BLOCK' || contentRoot('MPRR-V1-10-MUTATION-RESULT-V1', base) !== testRoot) throw new Error('corpus row mismatch');
  });
  const { corpusRoot, ...base } = corpus;
  if (contentRoot('MPRR-V1-10-MUTATION-CORPUS-V1', base) !== corpusRoot || canonical(runReviewProtocolMutationCampaign()) !== canonical(corpus.cases)) throw new Error('corpus execution/root mismatch');
}

function verifyCandidate() {
  const repositoryRoot = process.cwd();
  const manifest = readJson(repositoryRoot, PATHS.manifest).value; validateManifest(repositoryRoot, manifest);
  const registry = readJson(repositoryRoot, PATHS.registry).value; validateRegistry(registry);
  const sourceIndex = readJson(repositoryRoot, PATHS.sourceIndex).value; validateSourceIndex(sourceIndex, manifest.sourceCommit);
  const crosswalk = readJson(repositoryRoot, PATHS.crosswalk).value; validateCrosswalk(registry, crosswalk);
  const corpus = readJson(repositoryRoot, PATHS.corpus).value; validateCorpus(corpus);
  const positive = evaluateReviewProtocolInput(makeReviewProtocolVector());
  if (positive.status !== 'ELIGIBLE-PLANNING-VECTOR-NOT-AUTHORITY' || positive.validatorCount !== 15 || positive.blockedCount !== 0) throw new Error('positive protocol vector failed');
  if (evaluateCurrentReviewProtocolState().status !== 'BLOCKED') throw new Error('current state must remain blocked');
  const reader = manifest.members.find((member) => member.role === 'MPRRV110-READER-A');
  if (!reader) throw new Error('Reader A not pinned');
  const base = { artifactClass: 'CROSS-CHECK-READER-A-NOT-INDEPENDENT-NOT-ACCEPTANCE', artifactId: 'MPRR-V1-10-READER-A-REPORT-2026-08-30-G1', closureCount: 0, corpusRoot: corpus.corpusRoot, findingCount: 17, packageContentRoot: manifest.packageContentRoot, positiveValidatorCount: 15, readerArtifactSha256: reader.sha256, result: 'PASS-CANDIDATE-NOT-ACCEPTED', sourceCommit: manifest.sourceCommit, sourceSetRoot: sourceIndex.sourceSetRoot };
  return { ...base, reportRoot: contentRoot('MPRR-V1-10-READER-A-REPORT-V1', base) };
}

function runPython() {
  const result = spawnSync('python3', ['scripts/verify-review-protocol-v1-10-candidate.py'], { cwd: process.cwd(), encoding: 'utf8' });
  if (result.status !== 0) throw new Error(`Reader B failed: ${(result.stderr || result.stdout).trim()}`);
  return JSON.parse(result.stdout);
}

function patchFor(logicalPath, content) {
  if (fs.existsSync(logicalPath)) throw new Error(`report already exists: ${logicalPath}`);
  const lines = content.endsWith('\n') ? content.slice(0, -1).split('\n') : content.split('\n');
  return `*** Add File: ${logicalPath}\n${lines.map((line) => `+${line}`).join('\n')}\n`;
}

function emitReports(reportA, reportB) {
  if (runGit(['status', '--porcelain=v1', '--untracked-files=all']).trim() !== '') throw new Error('--emit-patch requires clean worktree');
  process.stdout.write(`*** Begin Patch\n${patchFor(PATHS.reportA, pretty(reportA))}${patchFor(PATHS.reportB, pretty(reportB))}*** End Patch\n`);
}

function checkExisting(reportA, reportB) {
  const storedA = readJson(process.cwd(), PATHS.reportA).value; const storedB = readJson(process.cwd(), PATHS.reportB).value;
  if (canonical(storedA) !== canonical(reportA) || canonical(storedB) !== canonical(reportB)) throw new Error('stored report drift');
  return { status: 'PASS-EXISTING-PROTOCOL-V1-10-CANDIDATE-NOT-ACCEPTED', packageContentRoot: reportA.packageContentRoot, readerAReportRoot: reportA.reportRoot, readerBReportRoot: reportB.reportRoot };
}

const reportA = verifyCandidate(); const reportB = runPython();
for (const key of ['packageContentRoot', 'sourceSetRoot', 'corpusRoot', 'findingCount', 'positiveValidatorCount']) if (reportA[key] !== reportB[key]) throw new Error(`Reader fact mismatch: ${key}`);
if (process.argv.includes('--emit-patch')) emitReports(reportA, reportB);
else if (process.argv.includes('--check-existing')) process.stdout.write(`${JSON.stringify(checkExisting(reportA, reportB))}\n`);
else process.stdout.write(`${JSON.stringify({ status: 'PASS-PROTOCOL-V1-10-CANDIDATE-NOT-ACCEPTED', readerA: reportA, readerB: reportB })}\n`);

File: scripts/verify-review-protocol-v1-10-candidate.py
#!/usr/bin/env python3
"""Cross-runtime structural Reader for Review Protocol v1.10 Candidate."""

from __future__ import annotations

import hashlib
import json
import os
import stat
import subprocess
from pathlib import Path, PurePosixPath
from typing import Any


DATE = "2026-08-30"
PACKAGE_DIR = f"docs/planning/three-review-protocol-v1-10-g1-package-{DATE}"
PATHS = {
    "manifest": f"{PACKAGE_DIR}/normative-package-manifest.json",
    "registry": f"{PACKAGE_DIR}/normative-registry.json",
    "source": f"{PACKAGE_DIR}/frozen-source-index.json",
    "crosswalk": f"{PACKAGE_DIR}/closure-crosswalk.json",
    "corpus": f"{PACKAGE_DIR}/mutation-corpus.json",
}
FINDINGS = [f"MPRR-V19-IHR-F{index:03d}" for index in range(1, 18)]
VALIDATORS = [
    "VALIDATOR-PACKAGE", "VALIDATOR-FROZEN-SOURCES", "VALIDATOR-SCHEMAS", "VALIDATOR-CLOSURE", "VALIDATOR-SEMANTIC-ENTAILMENT",
    "VALIDATOR-PREDECESSOR-BEHAVIOR", "VALIDATOR-CAUSAL-TRACE", "VALIDATOR-APPOINTMENTS", "VALIDATOR-EXTERNAL-SIGNATURES", "VALIDATOR-SCANNERS",
    "VALIDATOR-REMOTE-PUBLIC", "VALIDATOR-CAS", "VALIDATOR-RECOVERY", "VALIDATOR-TIME-REVOCATION-FINALITY", "VALIDATOR-THREE-REVIEWS-AND-HUMAN-APPROVAL",
]
ROLES = ["ROLE-PRODUCER-01", "ROLE-REVIEWER-01", "ROLE-REVIEWER-02", "ROLE-REVIEWER-03", "ROLE-RECONCILER-01", "ROLE-APPROVER-01", "ROLE-PERMIT-ISSUER-01"]
ALLOWED_PREFIXES = ("docs/planning/", "scripts/", "tests/")
SAFE_INTEGER_MAX = 9_007_199_254_740_991


def fail(message: str) -> None:
    raise ValueError(message)


def pairs_to_object(pairs: list[tuple[str, Any]]) -> dict[str, Any]:
    result: dict[str, Any] = {}
    for key, value in pairs:
        if key in result:
            fail(f"duplicate JSON key: {key}")
        result[key] = value
    return result


def normalize(value: Any) -> Any:
    if value is None or isinstance(value, bool):
        return value
    if isinstance(value, str):
        if any(0xD800 <= ord(character) <= 0xDFFF for character in value):
            fail("surrogate code point forbidden")
        return value
    if isinstance(value, int) and not isinstance(value, bool):
        if abs(value) > SAFE_INTEGER_MAX:
            fail("unsafe integer")
        return value
    if isinstance(value, float):
        fail("float forbidden")
    if isinstance(value, list):
        return [normalize(item) for item in value]
    if isinstance(value, dict):
        return {key: normalize(value[key]) for key in sorted(value)}
    fail(f"unsupported canonical type: {type(value).__name__}")


def canonical(value: Any) -> str:
    return json.dumps(normalize(value), ensure_ascii=False, separators=(",", ":"), sort_keys=True)


def sha256_bytes(value: bytes) -> str:
    return hashlib.sha256(value).hexdigest()


def root(domain: str, value: Any) -> str:
    return sha256_bytes(f"{domain}\n{canonical(value)}".encode("utf-8"))


def exact(value: Any, keys: list[str], label: str) -> dict[str, Any]:
    if not isinstance(value, dict) or sorted(value.keys()) != sorted(keys):
        fail(f"{label}: exact keys mismatch")
    return value


def assert_path(logical_path: str) -> None:
    if not isinstance(logical_path, str) or not logical_path.startswith(ALLOWED_PREFIXES) or logical_path.startswith("/") or "\\" in logical_path or "//" in logical_path:
        fail("invalid repository path")
    if any(part in ("", ".", "..") for part in logical_path.split("/")) or str(PurePosixPath(logical_path)) != logical_path:
        fail("non-canonical repository path")


def read_current(repository: Path, logical_path: str, maximum: int) -> bytes:
    assert_path(logical_path)
    root_path = repository.resolve(strict=True)
    cursor = root_path
    for index, component in enumerate(logical_path.split("/")):
        cursor = cursor / component
        observed = os.lstat(cursor)
        if stat.S_ISLNK(observed.st_mode):
            fail(f"symlink rejected: {logical_path}")
        if index < len(logical_path.split("/")) - 1 and not stat.S_ISDIR(observed.st_mode):
            fail("non-directory ancestor")
        if index == len(logical_path.split("/")) - 1 and not stat.S_ISREG(observed.st_mode):
            fail("non-regular file")
    resolved = cursor.resolve(strict=True)
    try:
        resolved.relative_to(root_path)
    except ValueError as error:
        raise ValueError("path escapes repository") from error
    before = os.stat(cursor, follow_symlinks=False)
    if before.st_nlink != 1 or before.st_size >= maximum:
        fail("hard-link or oversize rejected")
    descriptor = os.open(cursor, os.O_RDONLY | getattr(os, "O_NOFOLLOW", 0))
    try:
        opened = os.fstat(descriptor)
        if (opened.st_dev, opened.st_ino, opened.st_size) != (before.st_dev, before.st_ino, before.st_size):
            fail("file changed during open")
        with os.fdopen(os.dup(descriptor), "rb") as stream:
            data = stream.read()
        if len(data) != opened.st_size:
            fail("short read")
        return data
    finally:
        os.close(descriptor)


def read_git(repository: Path, commit: str, logical_path: str, maximum: int) -> tuple[bytes, int]:
    assert_path(logical_path)
    tree = subprocess.run(["git", "ls-tree", "-z", "--full-tree", commit, "--", logical_path], cwd=repository, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE).stdout
    entries = [entry for entry in tree.split(b"\0") if entry]
    if len(entries) != 1:
        fail(f"Git path count mismatch: {logical_path}")
    metadata, observed_path = entries[0].split(b"\t", 1)
    mode, kind, oid = metadata.decode("ascii").split(" ")
    if mode not in ("100644", "100755") or kind != "blob" or observed_path.decode("utf-8") != logical_path:
        fail(f"Git path is not regular blob: {logical_path}")
    data = subprocess.run(["git", "cat-file", "blob", oid], cwd=repository, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE).stdout
    if len(data) >= maximum:
        fail("Git blob oversize")
    return data, int(mode[-3:], 8)


def load(repository: Path, logical_path: str) -> dict[str, Any]:
    data = read_current(repository, logical_path, 25 * 1024 * 1024)
    value = json.loads(data.decode("utf-8"), object_pairs_hook=pairs_to_object)
    canonical(value)
    if not isinstance(value, dict):
        fail("JSON root must be object")
    return value


def validate_manifest(repository: Path, manifest: dict[str, Any]) -> None:
    exact(manifest, ["artifactClass", "artifactId", "generatedAt", "maxMemberBytesExclusive", "maxTotalBytesInclusive", "memberCount", "members", "packageContentRoot", "packageId", "repositoryVisibility", "schemaVersion", "sourceCommit", "totalBytes"], "manifest")
    if manifest["repositoryVisibility"] != "PUBLIC" or manifest["schemaVersion"] != "MPRR-V1-10-PACKAGE-MANIFEST-V1" or manifest["memberCount"] != 11 or len(manifest["members"]) != 11:
        fail("manifest invariant mismatch")
    paths: set[str] = set(); hashes: set[str] = set(); roles: set[str] = set(); total = 0
    for index, member in enumerate(manifest["members"]):
        exact(member, ["bytes", "logicalPath", "ordinal", "role", "sha256"], f"member[{index}]")
        if member["ordinal"] != index + 1 or member["logicalPath"] in paths or member["sha256"] in hashes or member["role"] in roles:
            fail("manifest uniqueness/order mismatch")
        paths.add(member["logicalPath"]); hashes.add(member["sha256"]); roles.add(member["role"])
        if member["logicalPath"].startswith(("scripts/", "tests/")):
            data, _ = read_git(repository, manifest["sourceCommit"], member["logicalPath"], manifest["maxMemberBytesExclusive"])
        else:
            data = read_current(repository, member["logicalPath"], manifest["maxMemberBytesExclusive"])
        if len(data) != member["bytes"] or sha256_bytes(data) != member["sha256"]:
            fail("manifest member drift")
        total += member["bytes"]
    if not any(member["role"] == "MPRRV110-B0-CORE-DEPENDENCY" and member["logicalPath"] == "scripts/b0-v8-core.mjs" for member in manifest["members"]):
        fail("manifest missing transitive B0 core dependency")
    if total != manifest["totalBytes"] or total > manifest["maxTotalBytesInclusive"]:
        fail("manifest byte budget mismatch")
    projection = {"members": manifest["members"], "packageId": manifest["packageId"], "sourceCommit": manifest["sourceCommit"]}
    if root("MPRR-V1-10-PACKAGE-CONTENT-V1", projection) != manifest["packageContentRoot"]:
        fail("manifest root mismatch")


def validate_registry(registry: dict[str, Any]) -> None:
    exact(registry, ["algorithmPolicy", "artifactClass", "artifactId", "controlRows", "currentState", "externalEvidence", "growthPolicy", "ownerModel", "receiptSchema", "reportPolicy", "repositoryVisibility", "roleSlots", "schemaVersion", "validatorIds"], "registry")
    if registry["repositoryVisibility"] != "PUBLIC" or registry["validatorIds"] != VALIDATORS or registry["roleSlots"] != ROLES:
        fail("registry denominator mismatch")
    if registry["ownerModel"] != {"ownerCount": 1, "primaryBackupModel": "REMOVED", "workOwner": "Tal"}:
        fail("owner model mismatch")
    current = registry["currentState"]
    if current["acceptance"] != 0 or current["authorityOutputs"] != 0 or current["b0"] != "ABSENT" or current["gate29"] != "BLOCKED" or current["developmentFreeze"] != "ACTIVE" or current["status"] != "BLOCKED":
        fail("current state overclaim")
    if registry["externalEvidence"]["externalClosureCount"] != 0 or registry["externalEvidence"]["operationalEvidencePresent"] is not False:
        fail("external evidence overclaim")
    if registry["algorithmPolicy"]["approvedAlgorithms"] != [] or registry["algorithmPolicy"]["keyGenerationPerformed"] is not False:
        fail("cryptographic policy overclaim")
    if registry["reportPolicy"]["writesAllowed"] is not False or registry["growthPolicy"]["globalBudgetPresent"] is not False:
        fail("report/growth policy overclaim")
    schema = registry["receiptSchema"]
    schema_base = {key: value for key, value in schema.items() if key != "schemaRoot"}
    if schema["additionalProperties"] is not False or root("MPRR-V1-10-CLOSED-SCHEMA-V1", schema_base) != schema["schemaRoot"]:
        fail("receipt schema root mismatch")
    if len(registry["controlRows"]) != 17 or [row["findingId"] for row in registry["controlRows"]] != FINDINGS:
        fail("control denominator mismatch")
    if any(row["localStatus"] != "IMPLEMENTED-CANDIDATE" or row["closureStatus"] != "OPEN-PENDING-INDEPENDENT-EVIDENCE" for row in registry["controlRows"]):
        fail("control closure overclaim")


def validate_source(repository: Path, source: dict[str, Any], commit: str) -> None:
    exact(source, ["artifactClass", "artifactId", "repositoryVisibility", "rows", "schemaVersion", "sourceCount", "sourceSetRoot"], "source")
    if source["sourceCount"] != 11 or len(source["rows"]) != 11:
        fail("source denominator mismatch")
    paths: set[str] = set(); hashes: set[str] = set()
    for index, row in enumerate(source["rows"]):
        exact(row, ["bytes", "logicalPath", "mode", "ordinal", "sha256"], f"source[{index}]")
        if row["ordinal"] != index + 1 or row["logicalPath"] in paths or row["sha256"] in hashes:
            fail("source uniqueness/order mismatch")
        paths.add(row["logicalPath"]); hashes.add(row["sha256"])
        data, mode = read_git(repository, commit, row["logicalPath"], 50 * 1024 * 1024)
        if len(data) != row["bytes"] or sha256_bytes(data) != row["sha256"] or mode != row["mode"]:
            fail("source drift")
    if not any(row["logicalPath"] == "scripts/b0-v8-core.mjs" for row in source["rows"]):
        fail("source index missing transitive B0 core dependency")
    if root("MPRR-V1-10-SOURCE-SET-V1", source["rows"]) != source["sourceSetRoot"]:
        fail("source root mismatch")


def validate_crosswalk(registry: dict[str, Any], crosswalk: dict[str, Any]) -> None:
    if crosswalk["findingCount"] != 17 or crosswalk["closureCount"] != 0 or crosswalk["inheritedFindingCount"] != 40 or crosswalk["inheritedIndependentMechanicalClosureCount"] != 1 or len(crosswalk["rows"]) != 17:
        fail("crosswalk denominator mismatch")
    for index, row in enumerate(crosswalk["rows"]):
        control = {key: value for key, value in row.items() if key != "ordinal"}
        if row["ordinal"] != index + 1 or canonical(control) != canonical(registry["controlRows"][index]):
            fail("crosswalk row mismatch")
    base = {key: value for key, value in crosswalk.items() if key != "crosswalkRoot"}
    if root("MPRR-V1-10-CROSSWALK-V1", base) != crosswalk["crosswalkRoot"]:
        fail("crosswalk root mismatch")


def validate_corpus(corpus: dict[str, Any]) -> None:
    if corpus["caseCount"] != 17 or corpus["blockedCount"] != 17 or len(corpus["cases"]) != 17:
        fail("corpus denominator mismatch")
    for index, row in enumerate(corpus["cases"]):
        base = {key: value for key, value in row.items() if key != "testRoot"}
        if row["ordinal"] != index + 1 or row["findingId"] != FINDINGS[index] or row["actual"] != "BLOCK" or root("MPRR-V1-10-MUTATION-RESULT-V1", base) != row["testRoot"]:
            fail("corpus row mismatch")
    base = {key: value for key, value in corpus.items() if key != "corpusRoot"}
    if root("MPRR-V1-10-MUTATION-CORPUS-V1", base) != corpus["corpusRoot"]:
        fail("corpus root mismatch")


def main() -> None:
    repository = Path.cwd()
    manifest = load(repository, PATHS["manifest"]); validate_manifest(repository, manifest)
    registry = load(repository, PATHS["registry"]); validate_registry(registry)
    source = load(repository, PATHS["source"]); validate_source(repository, source, manifest["sourceCommit"])
    crosswalk = load(repository, PATHS["crosswalk"]); validate_crosswalk(registry, crosswalk)
    corpus = load(repository, PATHS["corpus"]); validate_corpus(corpus)
    reader = next((member for member in manifest["members"] if member["role"] == "MPRRV110-READER-B"), None)
    if reader is None:
        fail("Reader B not pinned")
    base = {"artifactClass": "CROSS-RUNTIME-STRUCTURAL-READER-B-NOT-INDEPENDENT-NOT-ACCEPTANCE", "artifactId": "MPRR-V1-10-READER-B-REPORT-2026-08-30-G1", "closureCount": 0, "corpusRoot": corpus["corpusRoot"], "findingCount": 17, "packageContentRoot": manifest["packageContentRoot"], "positiveValidatorCount": 15, "readerArtifactSha256": reader["sha256"], "result": "PASS-STRUCTURAL-CANDIDATE-NOT-ACCEPTED", "sourceCommit": manifest["sourceCommit"], "sourceSetRoot": source["sourceSetRoot"]}
    report = {**base, "reportRoot": root("MPRR-V1-10-READER-B-REPORT-V1", base)}
    print(json.dumps(report, ensure_ascii=False, separators=(",", ":"), sort_keys=True))


if __name__ == "__main__":
    main()
